[CN] to_bytes and from_bytes crash without info about the error

I believe the error here is because of trying to convert an `Owned` resource but I am not actually sure. I would like info about what failed, but if that's tricky or the semantics are in flux just the source location of the failure would be helpful.

to_bytes.c:
```
typedef unsigned char uint8_t;
typedef struct s {
    uint8_t a;
} s;
/*@
predicate (map<u64,u8>) Array_u8 (pointer p, u64 l)
{
  take pv = each(u64 i; i >= 0u64 && i < l) {Owned<uint8_t>(array_shift<uint8_t>(p,i))};
  return pv;
}
@*/

#if 0
static void transmute_to_blob(void *p)
/*@
  requires take xsi = Array_u8(p, sizeof<s>);
  ensures take xso = Owned<s>(p);
  @*/
{
  /*@ from_bytes Owned<s>(p); @*/
}
#else

static void untransmute_to_blob(void *p)
/*@
  requires take xsi = Owned<s>(p);
  ensures take xso = Array_u8(p, sizeof<s>);
  @*/
{
  /*@ to_bytes Owned<s>(p); @*/
}
#endif
```
```
% cn verify to_bytes.c
to_bytes.c:30:7: warning: experimental keyword 'to_bytes' (use of experimental features is discouraged)
  /*@ to_bytes Owned<s>(p); @*/
      ^~~~~~~~ 
cn: internal error, uncaught exception:
    File "backend/cn/lib/check.ml", line 1866, characters 74-80: Assertion failed
    Raised at Cn__Check.check_expr.bytes_constraints in file "backend/cn/lib/check.ml", line 1866, characters 74-86
...
```
```
% cn verify to_bytes.c
to_bytes.c:20:7: warning: experimental keyword 'from_bytes' (use of experimental features is discouraged)
  /*@ from_bytes Owned<s>(p); @*/
      ^~~~~~~~~~ 
cn: internal error, uncaught exception:
    File "backend/cn/lib/check.ml", line 1866, characters 74-80: Assertion failed
    Raised at Cn__Check.check_expr.bytes_constraints in file "backend/cn/lib/check.ml", line 1866, characters 74-86
...
```

--- a/backend/cn/lib/check.ml
+++ b/backend/cn/lib/check.ml
@@ -1881,7 +1881,8 @@ let rec check_expr labels (e : BT.t Mu.expr) (k : IT.t -> unit m) : unit m =
          (* FIXME this hard codes big endianness but this should be switchable *)
          let here = Locations.other __LOC__ in
          match ct with
-         | Sctypes.Void | Array (_, _) | Struct _ | Function (_, _, _) -> assert false
+         | Sctypes.Void | Array (_, _) | Struct _ | Function (_, _, _) ->
+           fail (fun _ -> { loc; msg = Unsupported_byte_conv_ct ct })
          | Integer it ->
            let bt = IT.get_bt value in
            let lhs = value in
@@ -1895,7 +1896,7 @@ let rec check_expr labels (e : BT.t Mu.expr) (k : IT.t -> unit m) : unit m =
              in
              List.fold_left (fun x y -> IT.add_ (x, y) here) b bytes
            in
-           eq_ (lhs, rhs) here
+           return (eq_ (lhs, rhs) here)
          | Pointer _ ->
            (* FIXME this totally ignores provenances *)
            let bt = WellTyped.default_quantifier_bt in
@@ -1910,7 +1911,7 @@ let rec check_expr labels (e : BT.t Mu.expr) (k : IT.t -> unit m) : unit m =
              in
              List.fold_left (fun x y -> IT.add_ (x, y) here) b bytes
            in
-           eq_ (lhs, rhs) here
+           return (eq_ (lhs, rhs) here)
        in
        let@ () = WellTyped.ensure_base_type loc ~expect Unit in
        let aux loc stmt =
@@ -1938,7 +1939,9 @@ let rec check_expr labels (e : BT.t Mu.expr) (k : IT.t -> unit m) : unit m =
            let@ () = add_r loc (Q (bytes_qpred q_sym ct pointer init), O byte_arr) in
            (match init with
             | Uninit -> add_c loc (LC.T (IT.eq_ (byte_arr, default_ map_bt here) here))
-            | Init -> add_c loc (LC.T (bytes_constraints ~value ~byte_arr ct)))
+            | Init ->
+              let@ constr = bytes_constraints ~value ~byte_arr ct in
+              add_c loc (LC.T constr))
          | To_from_bytes (From, { name = Owned (ct, init); pointer; _ }) ->
            let@ () = WellTyped.owned_ct_ok loc (ct, init) in
            let@ pointer = WellTyped.infer_term pointer in
@@ -1961,7 +1964,9 @@ let rec check_expr labels (e : BT.t Mu.expr) (k : IT.t -> unit m) : unit m =
            in
            (match init with
             | Uninit -> add_c loc (LC.T (IT.eq_ (value, default_ value_bt here) here))
-            | Init -> add_c loc (LC.T (bytes_constraints ~value ~byte_arr ct)))
+            | Init ->
+              let@ constr = bytes_constraints ~value ~byte_arr ct in
+              add_c loc (LC.T constr))
          | Have lc ->
            let@ _lc = WellTyped.logical_constraint loc lc in
            fail (fun _ ->

--- a/backend/cn/lib/typeErrors.ml
+++ b/backend/cn/lib/typeErrors.ml
@@ -192,6 +192,7 @@ type message =
         orig_loc : Locations.t
       }
   | Requires_after_ensures of { ens_loc : Locations.t }
+  | Unsupported_byte_conv_ct of Sctypes.ctype
 
 type t =
   { loc : Locations.t;
@@ -541,6 +542,11 @@ let pp_message = function
     let head, pos = Locations.head_pos_of_location ens_loc in
     let descr = Some (!^"ensures clause at" ^^^ !^head ^/^ !^pos) in
     { short; descr; state = None }
+  | Unsupported_byte_conv_ct ct ->
+    let short =
+      !^"Cannot (yet) convert value of" ^^^ Sctypes.pp ct ^^^ !^"to/from bytes"
+    in
+    { short; descr = None; state = None }
 
 
 (** Convert a possibly-relative filepath into an absolute one. *)

--- a/backend/cn/lib/typeErrors.mli
+++ b/backend/cn/lib/typeErrors.mli
@@ -131,6 +131,7 @@ type message =
         orig_loc : Locations.t
       }
   | Requires_after_ensures of { ens_loc : Locations.t }
+  | Unsupported_byte_conv_ct of Sctypes.ctype
 
 type t =
   { loc : Locations.t;

--- a/tests/cn/to_from_bytes_struct.error.c
+++ b/tests/cn/to_from_bytes_struct.error.c
@@ -1,0 +1,32 @@
+typedef unsigned char uint8_t;
+typedef struct s {
+    uint8_t a;
+} s;
+/*@
+predicate (map<u64,u8>) Array_u8 (pointer p, u64 l)
+{
+  take pv = each(u64 i; i >= 0u64 && i < l) {RW<uint8_t>(array_shift<uint8_t>(p,i))};
+  return pv;
+}
+@*/
+
+#if 0
+static void transmute_to_blob(void *p)
+/*@
+  requires take xsi = Array_u8(p, sizeof<s>);
+  ensures take xso = RW<s>p);
+  @*/
+{
+  /*@ from_bytes RW<s>(p); @*/
+}
+#else
+
+static void untransmute_to_blob(void *p)
+/*@
+  requires take xsi = RW<s>(p);
+  ensures take xso = Array_u8(p, sizeof<s>);
+  @*/
+{
+  /*@ to_bytes RW<s>(p); @*/
+}
+#endif

--- a/tests/cn/to_from_bytes_struct.error.c.verify
+++ b/tests/cn/to_from_bytes_struct.error.c.verify
@@ -1,0 +1,8 @@
+return code: 1
+tests/cn/to_from_bytes_struct.error.c:30:7: warning: experimental keyword 'to_bytes' (use of experimental features is discouraged)
+  /*@ to_bytes RW<s>(p); @*/
+      ^~~~~~~~ 
+[1/1]: untransmute_to_blob -- fail
+tests/cn/to_from_bytes_struct.error.c:30:6: error: Cannot (yet) convert value of struct s to/from bytes
+  /*@ to_bytes RW<s>(p); @*/
+     ^~~~~~~~~~~~~~~~~~~~~ 

